### PR TITLE
fix(core): clear resolver cache for project files dev mode

### PR DESCRIPTION
**Pull request by pimlie**
## Types of changes

- [x] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

Partial fix for #6232 (that if serverMiddleware is provided as string then on dev server restart the changes in the middleware are not picked up).

I think this also fixes the issue I have during module development which caused me to having to restart the dev server all the time because changed werent picked up.

Not sure how to add a test for this.

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.

### CODE REVIEW

File: packages/core/src/resolver.js [hunk 1]
```diff
@@ -138,9 +138,17 @@ export default class Resolver {
       lastError = e
     }
 
+    const isExternal = resolvedPath.includes('/node_modules/')
+
+    // in dev mode make sure to clear the require cache so after
+    // a dev server restart any changed file is reloaded
+    if (this.options.dev && !isExternal && require.cache[resolvedPath]) {
+      delete require.cache[resolvedPath]
```
Comment (pi0) [inline packages/core/src/resolver.js]: We have a util (in cjs.js) to recursively cleanup required cache

File: packages/core/src/resolver.js [hunk 2]
```diff
@@ -138,9 +138,17 @@ export default class Resolver {
       lastError = e
     }
 
+    const isExternal = resolvedPath.includes('/node_modules/')
```
Comment (pi0) [inline packages/core/src/resolver.js]: Maybe also nice to export as a utility function

File: packages/core/src/resolver.js (hunk 1 above)
Comment (pimlie) [reply packages/core/src/resolver.js]: Ah yes, missed that. Using that now

File: packages/utils/src/cjs.js [hunk 3]
```diff
@@ -1,6 +1,10 @@
+export function isExternalDependency (id) {
+  return id.includes('/node_modules/')
+}
+
 export function clearRequireCache (id) {
   const entry = require.cache[id]
-  if (!entry || id.includes('node_modules')) {
+  if (!entry || isExternalDependency(id)) {
```
Comment (pimlie) [inline packages/utils/src/cjs.js]: not sure if it was deliberate to use `node_modules` without separators here?

**Comment by kevinmarrec** [comment]
@pi0 Any idea is it wouldn't cause issue to TypeScript runtime ?

File: packages/utils/src/cjs.js (hunk 3 above)
Comment (pi0) [reply packages/utils/src/cjs.js]: Indeed we can use path sep (don't forgot to include both \\ and / in regex ;))

**Comment by pi0** [comment]
@kevinmarrec Good point. We can try to ensure. I remember there was something wrong. But can try :)

**Comment by kevinmarrec** [comment]
Yeah I don't remember well what was going on around `ts-node` require cache.. It may be not an issue anymore 🤷‍♂️

File: packages/utils/src/cjs.js [hunk 4]
```diff
@@ -1,6 +1,12 @@
+import { sep } from 'path'
+
+export function isExternalDependency (id) {
+  return id.includes(`${sep}node_modules${sep}`)
```
Comment (pi0) [inline packages/utils/src/cjs.js]: I think it should cover **both** separators. A windows user can also call require with slash :)

**Review by pi0 — COMMENTED** [review]
-

**Review by pi0 — APPROVED** [review]
LGTM! It may (rare) be possible we introduce some perf penalties as we clear cache (so more CPU and Disk read) but from DX perspective seems a good idea.

**Review by aldarund — APPROVED** [review]

**Review by atinux — APPROVED** [review]

**Pull request merged**